src/Algebra/Graph/Undirected.hs

```diff
@@ -0,0 +1,913 @@
+/canonical graph representation/ and satisfies all axioms of algebraic graphs:
+
+    * 'overlay' is commutative and associative:
+
+        >       x + y == y + x
+        > x + (y + z) == (x + y) + z
+
+    * 'connect' is associative, commutative and has 'empty' as the identity:
+
+        >   x * empty == x
+        >   empty * x == x
+        >   x * y     == y * x
+        > x * (y * z) == (x * y) * z
+
+    * 'connect' distributes over 'overlay':
+
+        > x * (y + z) == x * y + x * z
+        > (x + y) * z == x * z + y * z
+
+    * 'connect' can be decomposed:
+
+        > x * y * z == x * y + x * z + y * z
+
+The following useful theorems can be proved from the above set of axioms.
+
+    * 'overlay' has 'empty' as the identity and is idempotent:
+
+        >   x + empty == x
+        >   empty + x == x
+        >       x + x == x
+
+    * Absorption and saturation of 'connect':
+
+        > x * y + x + y == x * y
+        >     x * x * x == x * x
+
+When specifying the time and memory complexity of graph algorithms, /n/ will
+denote the number of vertices in the graph, /m/ will denote the number of
+edges in the graph, and /s/ will denote the /size/ of the corresponding
+'Graph' expression. For example, if @g@ is an Undirected 'Graph' then /n/, /m/ and /s/ can
+be computed as follows:
+
+@n == 'vertexCount' g
+m == 'edgeCount' g
+s == 'size' g@
+
+Note that 'size' counts all leaves of the expression:
+
+@'vertexCount' 'empty'           == 0
+'size'        'empty'           == 1
+'vertexCount' ('vertex' x)      == 1
+'size'        ('vertex' x)      == 1
+'vertexCount' ('empty' + 'empty') == 0
+'size'        ('empty' + 'empty') == 2@
+
+Converting an Undirected 'Graph' to the corresponding 'AM.AdjacencyMap' takes /O(s + m * log(m))/
+time and /O(s + m)/ memory. This is also the complexity of the graph equality
+test, because it is currently implemented by converting graph expressions to
+canonical representations based on adjacency maps.
+
+The total order on graphs is defined using /size-lexicographic/ comparison:
+
+* Compare the number of vertices. In case of a tie, continue.
+* Compare the sets of vertices. In case of a tie, continue.
+* Compare the number of edges. In case of a tie, continue.
+* Compare the sets of edges.
+
+Here are a few examples:
+
+@'vertex' 1 < 'vertex' 2
+'vertex' 3 < 'edge' 1 2
+'vertex' 1 < 'edge' 1 1
+'edge' 1 1 < 'edge' 1 2
+'edge' 1 2 < 'edge' 1 1 + 'edge' 2 2
+'edge' 1 2 < 'edge' 1 3
+'edge' 1 2 == 'edge' 2 1
+
+Note that the resulting order refines the 'isSubgraphOf' relation and is
+compatible with 'overlay' and 'connect' operations:
+
+@'isSubgraphOf' x y ==> x <= y@
+
+@'empty' <= x
+x     <= x + y
+x + y <= x * y@
+-}
+newtype Graph a = UG { toGraph :: G.Graph a }
+             deriving (Generic, NFData)
+
+instance (Show a, Ord a) => Show (Graph a) where
+  show = show . toSymmetricRelation
+
+{- See Note [Functions for rewrite rules] in 'Algebra.Graph' -}
```

Comment:
@nobrakal Do you think we need to use any rewrite rules in this module? We simply wrap `Graph` into a `newtype` and most of the functions below are defined simply via `coerce`. I have a feeling that the rewrite rules from your module may be sufficient, in which case this module could be simplified by removing all rewrite rules. What are your thoughts?

Reply:
Indeed, you do not need to use rules. Express `bindR` using `G.Graph`'s `bind`.
Also inline everything (or at least, remove (now) useless phase control for `bindR`, `foldg` `overlays`...), since we need to show to the compiler that fusion is possible !

Reply:
@nobrakal Thanks! 

@bolt12 Could you follow Alexandre's suggestions?

Reply:
@snowleopard Sure! :-) Just to make sure: I need to remove the rules, rewrite bind, and inline everything, right? I didn't understand the remove (now) part very well @nobrakal

Reply:
Sorry if my comment was not clear. We need to:

* Remove all rules from `Algebra.Graph.Undirected` concerning functions that are just synonym of functions from `Algebra.Graph` (since their rules are already living in `Algebra.Graph`).
* Remove phase control (the [X] part in `INLINE` pragma) for functions that are just synonym of functions from `Algebra.Graph`. They are not needed anymore.
* The goal is to use rules from `Algebra.Graph`, thus we need to express all possible functions using functions from `Algebra.Graph`. This apply particularly to `bindR`.

This should (I hope) make things work :)

Reply:
Yes, thank you very much!